AX: consider z-index when computing ignored when a modal is present

hoffmanjoshua opened the pull request:
#### a2ad036263e755b05cf39684ac4d2b76c6877639
<pre>
AX: consider z-index when computing ignored when a modal is present
<a href="https://bugs.webkit.org/show_bug.cgi?id=295404">https://bugs.webkit.org/show_bug.cgi?id=295404</a>
<a href="https://rdar.apple.com/154886931">rdar://154886931</a>

Reviewed by Tyler Wilcock.

There can be authoring errors when using aria-modal, and frequently, absolutely positioned
content that should be floating above the modal is excluded from accessibility because of
our modal heuristics. We can be more forgiving by also checking if content has a higher
z-index than the modal, indicating that it should be accessible.

* LayoutTests/accessibility/element-outside-modal-with-zindex-expected.txt: Added.
* LayoutTests/accessibility/element-outside-modal-with-zindex.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::nearestRendererFromNode):
(WebCore::zIndexFromRenderer):
(WebCore::AccessibilityObject::ignoredFromModalPresence const):
(WebCore::AccessibilityObject::rendererOrNearestAncestor const):
* Source/WebCore/accessibility/AccessibilityObject.h:

Canonical link: <a href="https://commits.webkit.org/297016@main">https://commits.webkit.org/297016@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b79078d38f0d9ca8e7f2e9d728a32087b051a581

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110255 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29914 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20346 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116279 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60503 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112218 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30593 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38501 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83825 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99285 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64273 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23763 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17421 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60073 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93778 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17479 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119069 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37295 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27651 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92799 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37667 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95555 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92625 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23610 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37620 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15351 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33184 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37189 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42660 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36851 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40191 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38560 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -2433,6 +2434,17 @@ bool AccessibilityObject::isModalNode() const
+        int zIndex = layer->zIndex();
+        if (zIndex)
+            return zIndex;
+    }
```
I think this can be shortened to:

```cpp
for (CheckedPtr layer = renderLayer; layer; layer = layer->parent()) {
    if (int zIndex = layer->zIndex())
        return zIndex;
}
```

twilco commented on LayoutTests/accessibility/element-outside-modal-with-zindex.html:
```diff
@@ -0,0 +1,39 @@
+
+<div id="text2" style="position: absolute; width: 200px;">This is content floating beneath the modal.</div>
+
+
```
Extra newline here I think

twilco commented on LayoutTests/accessibility/element-outside-modal-with-zindex.html:
```diff
@@ -0,0 +1,39 @@
+        finishJSTest();
+    }, 0);
+}
+
```
Extra newline here I think

twilco commented on LayoutTests/accessibility/element-outside-modal-with-zindex.html:
```diff
@@ -0,0 +1,39 @@
+}
+
+</script>
+
```
Extra newline here I think

minorninth commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -2465,6 +2477,16 @@ bool AccessibilityObject::ignoredFromModalPresence() const
+    // If this element has a higher z-index than the active modal, also don't ignore it.
+    if (CheckedPtr renderer = this->renderer()) {
+        if (CheckedPtr modalRenderer = modalNode->renderer()) {
+            int thisZIndex = zIndexFromRenderLayer(renderer->enclosingLayer());
```
Tiny nit, but I feel like it'd be cleaner to have zIndexFromRenderer and have the helper call enclosingLayer().

minorninth commented on LayoutTests/accessibility/element-outside-modal-with-zindex.html:
```diff
@@ -0,0 +1,39 @@
+    <button id="modalButton" aria-label="Button"></button>
+</div>
+
+<div id="text1" style="z-index: 50; width: 200px; position: absolute;">This is content floating above the modal.</div>
```
It'd be nice to test that it gets the z-index from an ancestor and not just one explicitly set on that exact element. How about something like:

<div style="z-index: 50; position: absolute">
  <div style="position: relative; left:  10px;">
    <div style="position: relative; top: 10px;">
      <div id="text1">
        Check this element
      </div>
...

hoffmanjoshua replied:
Added a test case for this!

minorninth reviewed (commented):
lgtm

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -2465,6 +2476,16 @@ bool AccessibilityObject::ignoredFromModalPresence() const
     }
 
+    // If this element has a higher z-index than the active modal, also don't ignore it.
+    if (CheckedPtr renderer = this->rendererOrNearestAncestor()) {
```
Should we have a test case for display:contents?

hoffmanjoshua replied:
Updated with an additional test case for this!

twilco replied:
Thank you 🙏

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -2465,6 +2476,16 @@ bool AccessibilityObject::ignoredFromModalPresence() const
 
+    // If this element has a higher z-index than the active modal, also don't ignore it.
+    if (CheckedPtr renderer = this->rendererOrNearestAncestor()) {
+        if (CheckedPtr modalRenderer = modalNode->renderer()) {
```
Should we handle the case where the modal element has display:contents?

hoffmanjoshua replied:
Fixed

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -2898,6 +2919,15 @@ Element* AccessibilityObject::element() const
+{
+    CheckedPtr renderer = this->renderer();
+    for (RefPtr ancestor = this; ancestor && !renderer; ancestor = ancestor->parentObject())
+        renderer = ancestor->renderer();
```
Should we climb the renderer hierarchy rather than the accessibility hierarchy?

hoffmanjoshua replied:
(Discussed offline) We're going to start walking the DOM, since we are looking for a renderer. This will be cheaper than the AX tree walk.

twilco commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -2898,6 +2929,14 @@ Element* AccessibilityObject::element() const
+    if (RefPtr node = this->node())
+        return nearestRendererFromNode(*node);
+
+    return nullptr;
```
Could be written a bit more simply as:

```cpp
RefPtr node = this->node();
return node ? nearestRendererFromNode(*node) : nullptr;
```

hoffmanjoshua committed:
AX: consider z-index when computing ignored when a modal is present
https://bugs.webkit.org/show_bug.cgi?id=295404
rdar://154886931

Reviewed by Tyler Wilcock.

There can be authoring errors when using aria-modal, and frequently, absolutely positioned
content that should be floating above the modal is excluded from accessibility because of
our modal heuristics. We can be more forgiving by also checking if content has a higher
z-index than the modal, indicating that it should be accessible.

* LayoutTests/accessibility/element-outside-modal-with-zindex-expected.txt: Added.
* LayoutTests/accessibility/element-outside-modal-with-zindex.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::nearestRendererFromNode):
(WebCore::zIndexFromRenderer):
(WebCore::AccessibilityObject::ignoredFromModalPresence const):
(WebCore::AccessibilityObject::rendererOrNearestAncestor const):
* Source/WebCore/accessibility/AccessibilityObject.h:

Canonical link: https://commits.webkit.org/297016@main
